File: planet/services/ai-orchestrator/src/providers/gemini.ts
import type { AnalyzedContribution } from "@planet/shared-types";
import { ProviderUnavailableError, type AnalyzeRequest, type LLMProvider, type NarrateRequest, type NarrateResult } from "./base.js";
import { ANALYSIS_SYSTEM_PROMPT, NARRATIVE_SYSTEM_PROMPT } from "./openai.js";

/** Gemini adapter — live only when GEMINI_API_KEY is configured. */
export class GeminiProvider implements LLMProvider {
  name = "gemini";
  private apiKey = process.env["GEMINI_API_KEY"] ?? "";
  private model = process.env["GEMINI_MODEL"] ?? "gemini-1.5-flash";

  isLive(): boolean {
    return this.apiKey.length > 0;
  }

  private async generate(system: string, user: string, jsonMode: boolean): Promise<string> {
    const res = await fetch(
      `https://generativelanguage.googleapis.com/v1beta/models/${this.model}:generateContent?key=${this.apiKey}`,
      {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({
          systemInstruction: { parts: [{ text: system }] },
          contents: [{ role: "user", parts: [{ text: user }] }],
          generationConfig: {
            temperature: jsonMode ? 0.2 : 0.7,
            responseMimeType: jsonMode ? "application/json" : "text/plain",
          },
        }),
      },
    );
    if (!res.ok) throw new Error(`gemini_http_${res.status}`);
    const data = (await res.json()) as { candidates: { content: { parts: { text: string }[] } }[] };
    return data.candidates[0]?.content.parts.map((p) => p.text).join("") ?? "";
  }

  async analyze(req: AnalyzeRequest): Promise<AnalyzedContribution> {
    if (!this.isLive()) throw new ProviderUnavailableError(this.name);
    const text = await this.generate(ANALYSIS_SYSTEM_PROMPT, `category: ${req.category}\ntext: ${req.text}`, true);
    const parsed = JSON.parse(text) as Omit<AnalyzedContribution, "category" | "sandbox">;
    return { ...parsed, category: req.category, sandbox: false };
  }

  async narrate(req: NarrateRequest): Promise<NarrateResult> {
    if (!this.isLive()) throw new ProviderUnavailableError(this.name);
    const text = await this.generate(NARRATIVE_SYSTEM_PROMPT(req.locale), JSON.stringify(req.facts), false);
    return { text, llmGenerated: true };
  }
}

File: services/ai-orchestrator/app/providers/gemini.py
"""Gemini adapter (used only when GEMINI_API_KEY is configured)."""
from __future__ import annotations

import json
import os
from typing import Any

import httpx

from ..narrative import template_narrative
from ..schemas import StructuredContribution
from .prompts import PARSE_SYSTEM, NARRATE_SYSTEM, events_block

MODEL = os.getenv("GEMINI_MODEL", "gemini-1.5-flash")


class GeminiProvider:
    name = "gemini"

    def configured(self) -> bool:
        return bool(os.getenv("GEMINI_API_KEY"))

    def parse_contribution(self, text: str, locale: str, category: str | None) -> StructuredContribution:
        from .mock import MockProvider
        fallback = MockProvider().parse_contribution(text, locale, category)
        try:
            raw = self._chat(PARSE_SYSTEM, json.dumps({
                "text": text, "locale": locale, "categoryHint": category,
                "draft": fallback.model_dump(),
            }, ensure_ascii=False))
            data = json.loads(raw)
            data["provider"] = self.name
            data["sandbox"] = False
            data.setdefault("traits", fallback.traits)
            data.setdefault("possibleBiomes", fallback.possibleBiomes)
            data.setdefault("risks", fallback.risks)
            return StructuredContribution(**data)
        except Exception:
            return fallback

    def narrate(self, events: list[dict[str, Any]], locale: str) -> str:
        try:
            return self._chat(NARRATE_SYSTEM, events_block(events, locale))
        except Exception:
            return template_narrative(events, locale)

    def _chat(self, system: str, user: str) -> str:
        url = (
            f"https://generativelanguage.googleapis.com/v1beta/models/{MODEL}:generateContent"
            f"?key={os.environ['GEMINI_API_KEY']}"
        )
        with httpx.Client(timeout=30) as client:
            res = client.post(url, json={
                "systemInstruction": {"parts": [{"text": system}]},
                "contents": [{"parts": [{"text": user}]}],
                "generationConfig": {
                    "responseMimeType": "application/json",
                    "maxOutputTokens": int(os.getenv("AI_MAX_TOKENS", "2048")),
                },
            })
            res.raise_for_status()
            return res.json()["candidates"][0]["content"]["parts"][0]["text"]
